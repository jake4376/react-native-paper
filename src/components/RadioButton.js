/* @flow */

import React, {
  PropTypes,
} from 'react';
import {
  StyleSheet,
} from 'react-native';
import color from 'color';
import Icon from './Icon';
import TouchableRipple from './TouchableRipple';
import withTheme from '../core/withTheme';
import type { Theme } from '../types/Theme';

type Props = {
  checked: boolean;
  disabled?: boolean;
  onPress?: Function;
  color?: string;
  theme: Theme;
}

const RadioButton = (props: Props) => {
  const {
    checked,
    disabled,
    onPress,
    theme,
  } = props;

  const radioColor = props.color || theme.colors.primary;

  let rippleColor, radioStyle;

  if (disabled) {
    rippleColor = 'rgba(0, 0, 0, .16)';
    radioStyle = { color: 'rgba(0, 0, 0, .26)' };
  } else {
    rippleColor = color(radioColor).clearer(0.32).rgbaString();
    if (checked) {
      radioStyle = { color: radioColor };
    }
  }

  return (
    <TouchableRipple
      {...props}
      borderless
      rippleColor={rippleColor}
      onPress={disabled ? undefined : onPress}
      style={styles.container}
    >
      <Icon
        name={checked ? 'radio-button-checked' : 'radio-button-unchecked'}
        size={24}
        style={[ styles.radio, radioStyle ]}
      />
    </TouchableRipple>
  );
};

RadioButton.propTypes = {
  checked: PropTypes.bool.isRequired,
  disabled: PropTypes.bool,
  onPress: PropTypes.func,
  color: PropTypes.string,
  theme: PropTypes.object.isRequired,
};

const styles = StyleSheet.create({
  container: {
    borderRadius: 18,
  },

  radio: {
    /* FIXME: using opacity doesn't work properly with TouchableHighlight */
    color: 'rgba(0, 0, 0, .54)',
    margin: 6,
  },
});

export default withTheme(RadioButton);
