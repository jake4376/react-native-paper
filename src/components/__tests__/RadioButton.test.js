import React from 'react';
import renderer from 'react-test-renderer';
import { RadioButtonContext } from '../RadioButtonGroup';
import RadioButton from '../RadioButton';

describe('RadioButton', () => {
  afterEach(() => jest.resetModules());

  describe('on default platform', () => {
    beforeAll(() => {
      jest.doMock('Platform', () => ({ select: objs => objs.default }));
    });

    it('renders properly', () => {
      const tree = renderer.create(<RadioButton value="first" />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('on ios platform', () => {
    beforeAll(() => {
      jest.doMock('Platform', () => ({ select: objs => objs.ios }));
    });

    it('renders properly', () => {
      const tree = renderer.create(<RadioButton value="first" />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('when RadioButton is wrapped by RadioButtonContext.Provider', () => {
    it('renders properly', () => {
      const tree = renderer
        .create(
          <RadioButtonContext.Provider value="first" onValueChange={() => {}}>
            <RadioButton value="first" />
          </RadioButtonContext.Provider>
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
