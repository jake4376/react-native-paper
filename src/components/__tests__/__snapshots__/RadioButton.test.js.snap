// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton on default platform renders properly 1`] = `
<View
  accessibilityRole="button"
  accessibilityStates={Array []}
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 18,
      },
    ]
  }
>
  <View
    style={
      Object {
        "borderColor": "rgba(0, 0, 0, 0.54)",
        "borderRadius": 10,
        "borderWidth": 2,
        "height": 20,
        "margin": 8,
        "width": 20,
      }
    }
  />
</View>
`;

exports[`RadioButton on ios platform renders properly 1`] = `
<View
  accessibilityRole="button"
  accessibilityStates={Array []}
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 18,
        "padding": 6,
      },
    ]
  }
>
  <View
    style={
      Object {
        "opacity": 0,
      }
    }
  >
    <Text
      accessibilityElementsHidden={true}
      allowFontScaling={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "#03dac4",
            "fontSize": 24,
          },
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleX": 1,
                },
              ],
            },
            Object {
              "backgroundColor": "transparent",
            },
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`RadioButton when RadioButton is wrapped by RadioButtonContext.Provider renders properly 1`] = `
<View
  accessibilityRole="button"
  accessibilityStates={Array []}
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 18,
        "padding": 6,
      },
    ]
  }
>
  <View
    style={
      Object {
        "opacity": 0,
      }
    }
  >
    <Text
      accessibilityElementsHidden={true}
      allowFontScaling={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "#03dac4",
            "fontSize": 24,
          },
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleX": 1,
                },
              ],
            },
            Object {
              "backgroundColor": "transparent",
            },
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;
